File: solutions/2022/01/go/main.go
package main

import (
	"github.com/wrporter/advent-of-code/internal/common/convert"
	"github.com/wrporter/advent-of-code/internal/common/ints"
	"github.com/wrporter/advent-of-code/internal/common/solution"
	"sort"
	"strings"
)

func main() {
	s := Solution{AbstractSolution: solution.AbstractSolution{
		Solution: Solution{},
		Year:     2022,
		Day:      1,
	}}

	s.Run([]interface{}{}, []interface{}{})
}

type Solution struct {
	solution.AbstractSolution
}

func (s Solution) Part1(input string, args ...interface{}) interface{} {
	lists := strings.Split(input, "\n\n")
	max := 0

	for _, listString := range lists {
		total := getTotal(listString)
		max = ints.Max(max, total)
	}

	return max
}

func (s Solution) Part2(input string, args ...interface{}) interface{} {
	lists := strings.Split(input, "\n\n")
	var totals []int

	for _, listString := range lists {
		total := getTotal(listString)
		totals = append(totals, total)
	}

	sort.Sort(sort.Reverse(sort.IntSlice(totals)))

	return totals[0] + totals[1] + totals[2]
}

func getTotal(listString string) int {
	calorieStrings := strings.Split(listString, "\n")
	total := 0
	for _, calorieString := range calorieStrings {
		total += convert.StringToInt(calorieString)
	}
	return total
}

File: solutions/2022/01/rust/main.rs
use std::any::Any;
use std::fmt::Debug;
use crate::solution::{AbstractSolution, AnyDebug, Solution};

mod solution;
mod color;


impl AbstractSolution for Solution {
    fn year(&self) -> u32 { 2022 }
    fn day(&self) -> u32 { 1 }

    fn part_1(&self, input: &str) -> Box<dyn AnyDebug> {
        let max = input.split("\n\n")
            .map(|elf| elf.lines()
                .map(|item| item.parse::<u32>().unwrap()).sum::<u32>())
            .max()
            .unwrap();

        Box::new(max)
    }

    fn part_2(&self, input: &str) -> Box<dyn AnyDebug> {
        let mut elves: Vec<u32> = input.split("\n\n")
            .map(|elf| elf.lines().map(|item| item.parse::<u32>().unwrap()).sum())
            .collect();

        elves.sort_by(|a, b| b.cmp(a));
        let top3: u32 = elves.into_iter().take(3).sum();

        Box::new(top3)
    }
}

fn main() {
    let runner = solution::Runner::new();
    let s = Box::new(solution::Solution {});
    runner.run(s)
}
